frontend/src/utils/emails/template-parser.ts

```diff
@@ -0,0 +1,69 @@
+// A naive template parser for the .template email files
+
+type EmailTemplate = {
+  subject: string
+  body: string
+}
+
+async function readFile(filename: string): Promise<string> {
+  return fetch(filename).then((r) => r.text())
+}
+
+function replaceTemplateTagsUtil(
+  text: string,
+  startIndex: number,
+  data: string[],
+  accum: string
+): string {
+  if (startIndex >= text.length) {
+    if (data)
+      throw new Error('Mismatch between number of tags and the data provided')
+    return accum
+  }
+
+  const replaceIndex = text.indexOf('<data>', startIndex)
+  if (replaceIndex === -1) return accum + text.slice(startIndex)
+  const dataPoint = data.shift()
+
+  const mid = text.slice(startIndex, replaceIndex)
+
+  if (!dataPoint)
+    throw new Error('Mismatch between number of tags and the data provided')
+  return replaceTemplateTagsUtil(
+    text,
+    replaceIndex + 6,
+    data,
+    accum + mid + dataPoint
+  )
+}
+
+function replaceTemplateTags(text: string, data: string[]) {
+  return replaceTemplateTagsUtil(text, 0, data, '')
+}
+
+function extractBetweenTag(tagName: string, text: string) {
+  const openTag = `<${tagName}>`,
+    closeTag = `</${tagName}>`
+  const firstIndex = text.indexOf(openTag)
+  const lastIndex = text.indexOf(closeTag)
+
+  if (firstIndex === -1 || lastIndex === -1) {
+    throw new Error(`Invalid text provided -- could not parse ${tagName}.`)
+  }
+
+  return text.substring(firstIndex + tagName.length + 2, lastIndex).trim()
+}
+
+export default async function parseTemplate(
```

Comment:
Needing to read files/make network requests to generate a URL string does lead to some strange situations during testing:
```
const [url, setUrl] = useState('')
useEffect(() => {
  getMailUrl(['cl859@cornell.edu', 'ag759@cornell.edu'], 'matched', []).then((url) => setUrl(url!))
}, [])
```
to get `<a href={url}>` to work, because href can't be given a Promise directly. But it's possible this is less of an issue if email links aren't present on the page initially

Reply:
I think if we need to use fetch to get the templates from the public folder, then we will absolutely need Promises. A solution could just be to still display the link without any href attached to it, and then use react state + useEffect to dynamically add the mailto URLs. If this happens when the website loads, then this wouldn't be an issue.

Reply:
Hmmm... this feels like a lot of overhead for something is ultimately a long string with some substitution on the client side. Plus, this would be a lot of requests with the current structure if there's a link prepared for every grouping on the edit page and it needs to fetch every time.